### PR TITLE
[Driver][Linux] Delegate linking of sanitizer runtimes to underlying Clang

### PR DESCRIPTION
**Pull request by yln**
More principled fix of: https://bugs.swift.org/browse/SR-10152
Supersedes: https://github.com/apple/swift/pull/23505

### CODE REVIEW

**Comment by yln** [comment]
@swift-ci Please test Linux

File: lib/Driver/UnixToolChains.cpp [hunk 1]
```diff
@@ -293,23 +273,12 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
...
-      if (context.OI.SelectedSanitizers & SanitizerKind::Undefined)
-        addLinkSanitizerLibArgsForLinux(context.Args, Arguments, "ubsan", *this);
-
-      if (context.OI.SelectedSanitizers & SanitizerKind::Fuzzer)
-        addLinkRuntimeLib(context.Args, Arguments,
-                          sanitizerRuntimeLibName("fuzzer"));
-
-    }
+  // On Linux, delegate to Clang and let it figure out the correct linker
+  // options for sanitizers.
```
Comment (jrose-apple) [inline lib/Driver/UnixToolChains.cpp]: I think we can just do this everywhere, not just Linux! That'll help the few people on BSDs as well, even before we finish getting Windows up and running.

File: lib/Driver/UnixToolChains.cpp [hunk 2]
```diff
@@ -293,23 +273,12 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
...
+  if (getTriple().getOS() == llvm::Triple::Linux &&
+      job.getKind() == LinkKind::Executable && context.OI.SelectedSanitizers) {
+    Arguments.push_back(context.Args.MakeArgString(
+        "-fsanitize=" + getSanitizerList(context.OI.SelectedSanitizers)));
```
Comment (jrose-apple) [inline lib/Driver/UnixToolChains.cpp]: Hm. This works, but is passing through the contents of Swift's own OPT_sanitize_EQ good enough? Just in the interest of less code to maintain.

File: lib/Driver/UnixToolChains.cpp (hunk 1 above)
Comment (compnerd) [reply lib/Driver/UnixToolChains.cpp]: Yeah, this should be safe to do generally.  Windows is not running too far behind these days!

File: lib/Driver/UnixToolChains.cpp (hunk 2 above)
Comment (compnerd) [reply lib/Driver/UnixToolChains.cpp]: This seems much safer, and might be worth the extra complexity.

Comment (jrose-apple) [reply lib/Driver/UnixToolChains.cpp]: Okay, I can be overruled here!

Comment (yln) [reply lib/Driver/UnixToolChains.cpp]: I will provide a follow-up patch to reduce the maintenance cost a bit.

**Review by beccadax — APPROVED** [review]

**Comment by yln** [comment]
@swift-ci Please test Linux

**Comment by yln** [comment]
@swift-ci Please test Linux

File: lib/Driver/UnixToolChains.cpp [hunk 3]
```diff
@@ -293,22 +273,14 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
...
-        addLinkRuntimeLib(context.Args, Arguments,
-                          sanitizerRuntimeLibName("fuzzer"));
-
+  // Delegate to Clang for sanitizers. It will figure out the correct linker
+  // options.
+  if (job.getKind() == LinkKind::Executable && context.OI.SelectedSanitizers) {
+    Arguments.push_back(context.Args.MakeArgString(
+        "-fsanitize=" + getSanitizerList(context.OI.SelectedSanitizers)));
+    if (context.OI.SelectedSanitizers & SanitizerKind::Thread) {
+      Arguments.push_back("-lBlocksRuntime");
```
Comment (compnerd) [inline lib/Driver/UnixToolChains.cpp]: Does tsan itself use blocks?  Or is the blocks dependency for libdispatch?

Comment (yln) [reply lib/Driver/UnixToolChains.cpp]: The TSan runtime itself now (unfortunately) uses blocks.
I noticed that I started conflating different issues in this PR. I will try to pick them apart and provide separate PRs.

**Comment by yln** [comment]
@swift-ci Please test Linux
@swift-ci Please smoke test osx

**Comment by yln** [comment]
@swift-ci Please test Linux

**Comment by yln** [comment]
@swift-ci Please smoke test osx

**Comment by yln** [comment]
@swift-ci Please smoke test OS X platform

**Comment by yln** [comment]
@jrose-apple @compnerd Everything okay with this?

**Review by jrose-apple — APPROVED** [review]

**Pull request merged**